src/android/Geolocation.java

```diff
@@ -46,46 +48,101 @@ public class Geolocation extends CordovaPlugin implements OnLocationResultEventL
     @Override
     public void initialize(CordovaInterface cordova, CordovaWebView webView) {
         super.initialize(cordova, webView);
-        locationContexts = new SparseArray<LocationContext>();
-        fusedLocationClient = LocationServices.getFusedLocationProviderClient(cordova.getActivity());
+        locationContexts = new SparseArray<>();
+        if(checkGMS()) {
+            fusedLocationClientGms = LocationServices.getFusedLocationProviderClient(cordova.getActivity());
+        } else if (checkHMS()) {
+            fusedLocationProviderClientHms = com.huawei.hms.location.LocationServices.getFusedLocationProviderClient(cordova.getActivity());
+        }
     }
 
     @Override
     public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
-        if(!checkGooglePlayServicesAvailable(callbackContext)) {
-            return false;
-        }
+        this.cordova.getActivity().runOnUiThread(() -> {
+            try {
+                if ("getLocation".equals(action)) {
+                    flowGetLocationGmsOrHms(args, callbackContext);
+                } else if ("addWatch".equals(action)) {
+                    flowAddWatchGmsOrHms(args, callbackContext);
+                } else if ("clearWatch".equals(action)) {
+                    flowClearWatchGmsOrHms(args, callbackContext);
+                }
+            } catch (Exception ex) {
+                callbackContext.error(ex.getMessage());
+            }
+        });
 
-        if ("getLocation".equals(action)) {
-            int id = args.getString(3).hashCode();
-            LocationContext lc = new LocationContext(id, LocationContext.Type.RETRIEVAL, args, callbackContext, this);
-            locationContexts.put(id, lc);
+        return true;
+    }
 
+    /**
+     * Decide if getLocation will be executed from GMS services (Google Services) or HMS (Huawei Services)
+     *
+     * @param args arguments received from execute
+     * @param callbackContext the callbackContext
+     */
+    private void flowGetLocationGmsOrHms(JSONArray args, CallbackContext callbackContext) throws JSONException {
+        int id = args.getString(3).hashCode();
+
+        if (checkGMS()) {
```

Comment:
Most of the code inside the `checkGMS/HMS` if statements is identical besides one or two property/methods. Can't we reuse code as much as possible in order to avoid too much replication?